[CI/Build] Add linting for github actions workflows

russellb opened the pull request:
This PR introduces linting for github actions workflows to help prevent any
accidental introduction of errors in the future. It introduces this linting in
two ways:

1. A way to run the linting locally as part of `format.sh`.

2. Automatically run the linter in CI anytime github actions workflows are
   modified.

71be98bf github: Add an actionlint github workflow
a9785b56 github: Quote variables
2dcdd71d github: Handle filenames with special characters
fc050dd9 github: Update versions of various github actions
0956a417 format.sh: Avoid exit when no files changed
276b4088 format.sh: lint github ci configuration

commit 71be98bfa693baa0711e1343adc7917a8e93892f
Author: Russell Bryant <rbryant@redhat.com>
Date:   Mon Aug 26 13:39:01 2024 -0400

    github: Add an actionlint github workflow
    
    Add a new github actions workflow that will perform linting on github
    actions workflow changes. It will only run if github workflows are
    modified.
    
    Signed-off-by: Russell Bryant <rbryant@redhat.com>

commit a9785b56661f0d65c22ad6eede1053baf1038223
Author: Russell Bryant <rbryant@redhat.com>
Date:   Mon Aug 26 13:30:03 2024 -0400

    github: Quote variables
    
    Avoid accidental globbing or argument splitting by quoting the
    `$GITHUB_ENV` variable. In practice, this would probably never be a
    problem since github sets the value, but it follows best shell
    practices and makes the shell linter happy.
    
    Signed-off-by: Russell Bryant <rbryant@redhat.com>

commit 2dcdd71d9fee141f61a4e05e9b49c38a90954dfb
Author: Russell Bryant <rbryant@redhat.com>
Date:   Mon Aug 26 13:34:03 2024 -0400

    github: Handle filenames with special characters
    
    Resolve the following linter warning:
    
    ```
    .github/workflows/publish.yml:87:9: shellcheck reported issue in this script: SC2011:warning:2:14: Use 'find .. -print0 | xargs -0 ..' or 'find .. -exec .. +' to allow non-alphanumeric filenames [shellcheck]
    ```
    
    I don't expect this was a problem in practice, but it follows best
    practice as indicated by the linter.
    
    Signed-off-by: Russell Bryant <rbryant@redhat.com>

commit fc050dd964c3c898c6cd49868cd177ac664aeace
Author: Russell Bryant <rbryant@redhat.com>
Date:   Mon Aug 26 14:27:30 2024 -0400

    github: Update versions of various github actions
    
    These were raised by the action linter.
    
    Signed-off-by: Russell Bryant <rbryant@redhat.com>

commit 0956a41791051ab49483751acf9e7e21b743b5c6
Author: Russell Bryant <rbryant@redhat.com>
Date:   Tue Aug 27 10:08:31 2024 -0400

    format.sh: Avoid exit when no files changed
    
    This script is set to exit any time a command fails. In the case of
    this line, the script unexpectedly errors out on this check to see if
    any files have changed that are relevant for the `clang-format` tool.
    The `grep` specifically will return non-zero in this case.
    
    The workaround here is to catch the error and run `echo -e` instead,
    which we know will succeed and the end result is what we want (the
    list of changed files will be empty).
    
    This allows the script to continue to completion, even when
    `clang-format` has nothing to do.
    
    Signed-off-by: Russell Bryant <rbryant@redhat.com>

commit 276b4088688f13d311f1807de634e20165c35166
Author: Russell Bryant <rbryant@redhat.com>
Date:   Tue Aug 27 11:08:50 2024 -0400

    format.sh: lint github ci configuration
    
    Update `format.sh` to run the `actionlint` tool to check for errors in
    GitHub CI configuration. Both this script and the workflow to run this
    tool in CI run it using docker to ensure that the same version is in
    use in both places.
    
    Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
github: Quote variables

Avoid accidental globbing or argument splitting by quoting the
`$GITHUB_ENV` variable. In practice, this would probably never be a
problem since github sets the value, but it follows best shell
practices and makes the shell linter happy.

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
github: Handle filenames with special characters

Resolve the following linter warning:

```
.github/workflows/publish.yml:87:9: shellcheck reported issue in this script: SC2011:warning:2:14: Use 'find .. -print0 | xargs -0 ..' or 'find .. -exec .. +' to allow non-alphanumeric filenames [shellcheck]
```

I don't expect this was a problem in practice, but it follows best
practice as indicated by the linter.

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
github: Add an actionlint github workflow

Add a new github actions workflow that will perform linting on github
actions workflow changes. It will only run if github workflows are
modified.

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
github: Update versions of various github actions

These were raised by the action linter.

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb commented:
/ready

njhill commented:
Thanks @russellb! I wonder if we could incorporate this into the existing `format.sh` assuming it doesn't take long to run? I.e. rather than having to run `make` separately for this...

russellb committed:
format.sh: Avoid exit when no files changed

This script is set to exit any time a command fails. In the case of
this line, the script unexpectedly errors out on this check to see if
any files have changed that are relevant for the `clang-format` tool.
The `grep` specifically will return non-zero in this case.

The workaround here is to catch the error and run `echo -e` instead,
which we know will succeed and the end result is what we want (the
list of changed files will be empty).

This allows the script to continue to completion, even when
`clang-format` has nothing to do.

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
format.sh: lint github ci configuration

Update `format.sh` to run the `actionlint` tool to check for errors in
GitHub CI configuration. Both this script and the workflow to run this
tool in CI run it using docker to ensure that the same version is in
use in both places.

Signed-off-by: Russell Bryant <rbryant@redhat.com>

khluu commented on format.sh:
```diff
@@ -286,6 +286,29 @@ else
+    fi
+    # Ensure we use the same version of actionlint as CI
+    IMAGE="vllm/actionlint:latest"
+    ${DOCKER} build --tag "${IMAGE}" - < .github/workflows/actionlint.dockerfile
```
Can we use something like `flake8` instead of Docker for local formatting? Some people might not have Docker installed ...

russellb replied:
It will at least skip this step gracefully if it's not installed.

The alternative is that `go` needs to be installed. It's not a Python tool. https://github.com/rhysd/actionlint

russellb replied:
They have pre-built binaries, so I could have it auto-download that. I'll look at that now.

russellb replied:
@khluu I've got it using a prebuilt binary from a github release if it's not found, so there are no longer any dependencies for this to work

khluu replied:
Thanks!

russellb committed:
actionlint: Switch to prebuilt binary instead of docker

Previously `format.sh` and the github workflow used docker or podman
to run actionlint. This now will use an installed version if found,
and will otherwise download and run a prebuilt binary from a github
release.

Signed-off-by: Russell Bryant <rbryant@redhat.com>
